0018-connector-risk-mitigations/0018-connector-risk-mitigations.md

```diff
@@ -17,25 +17,28 @@ The main risk connectors face in Interledger is being unable to fulfill the inco
+
 ## Liquidity Exhaustion
 
-Attackers could tie up connectors' liquidity by preparing payments through them that the attacker knows will fail. For example, an attacker could prepare numerous payments to itself and then not fulfill any of them.
+Attackers could tie up connectors' liquidity by preparing payments through them that the attacker knows will fail. For example, an attacker could prepare numerous payments to itself and then not fulfill any of them. In the case of ILPv4, this would require a significant amount of payments to cause damage because of the nature of streaming micropayments.
 
 Possible mitigations include:
 
 * **Managing "Payment Bandwidth"** - Connectors should monitor the percentage of their total liquidity that is on hold at any given time for a given peer, customer, or destination prefix and may reject incoming payment requests if that party exceeds their allocated "payment bandwidth". Connectors may allocate less bandwidth to unknown or untrustworthy senders or receivers. Note that ILP Packets contain the destination account but not the source account, so connectors should apply this logic to the immediate peer or customer that they receive the payments from and possibly the destination account or prefix.
-* **Preferring Smaller Payments** - Smaller payments place a smaller percentage of the connector's liquidity on hold and thus each one carries less risk that the payment being prepared and then failing would create significant opportunity cost
+* **Preferring Smaller Payments** - Smaller payments place a smaller percentage of the connector's liquidity on hold and thus each one carries less risk that the payment being prepared and then failing would create significant opportunity cost - in ILPv4, this is the default.
 * **Blacklisting Senders and Receivers** - Connectors may refuse to facilitate payments from certain sources or to certain destinations if their rate of failed payments is unusually high
 
+In the worst case, during the fulfill phase a connector along the path fulfilling the payment could withhold the fulfill packet such that their corresponding connector would not receive a payment. The corresponding connector would at worst lose the unit amount being streamed in a single packet (i.e. 1 satoshi). The fulfilling connector would gain the unit amount and would put their reputation as a connector on the line and would miss out on future fee gains. Additionally, the fulfilling connector would have to close the payment channel between them and the corresponding connector, as the corresponding connector would choose not to facilitate payments through the channel after being cheated. This would result in an on-chain transaction which would make the economics for the attacker not worth it, as they could be making more by continuing to facilitate fees.
```

Comment:
> The fulfilling connector would gain the unit amount

This is not true. By withholding the fulfillment the upstream connector is denying themselves a payment (and all connectors downstream).

Reply:
What about in the case of a connector getting a fulfillment from a previous connector but not sending a fulfill to the next connector?

Reply:
Let's say you have connectors A, B, and C and the payment is going from A -> B -> C. Take the case of B. C passes the fulfillment to B, meaning that B now needs to pay. If B doesn't pass the fulfillment on to A, it's _B_ that looses the money. A doesn't care, because A hasn't paid anything yet.

Reply:
>  C passes the fulfillment to B, meaning that B now needs to pay. 
I need some clarity on this. So fulfillments don't imply payments which is why my description is incorrect?

Ok, let's say B passes the fulfillment on to A. But B hasn't paid C. Does A pay B? Does A know that C hasn't been paid and act accordingly? (and would that apply if we had another node, D as the receiver, and C as a connector)?

I just need to change this description to talk about not paying rather than withholding a fulfillment right?

Reply:
Each settlement agreement is handled on a bilateral basis. Once B has gotten the fulfillment from C, B owes C the money. If B screws C, that's C's problem. If B gives the fulfillment to A, B has met their side of the agreement with A and A owes B the money.

The not paying problem is simply the counterparty risk that is limited by the total value of the Interledger packets you're willing to forward for your counterparty before they send you the corresponding on-ledger money (or payment channel claims).